src/PythonExtensionGen.cpp

```diff
@@ -151,112 +123,125 @@ void PythonExtensionGen::compile(const Module &module) {
 /* Older Python versions don't set up PyMODINIT_FUNC correctly. */
 #if defined(_MSC_VER)
-#    define HALIDE_PYTHON_EXPORT __declspec(dllexport)
+#define HALIDE_PYTHON_EXPORT      __declspec(dllexport)
 #else
-#    define HALIDE_PYTHON_EXPORT __attribute__((visibility("default")))
+#define HALIDE_PYTHON_EXPORT      __attribute__((visibility("default")))
 #endif
 
-#ifdef __cplusplus
-extern "C" {
-#endif
+namespace {
 
-static
-#if !defined(_MSC_VER)
-__attribute__((unused))
-#endif
-int _convert_py_buffer_to_halide(
-        PyObject* pyobj, int dimensions, int flags,
-        halide_dimension_t* dim,  // array of >= size `dimensions`
-        halide_buffer_t* out, Py_buffer &buf, const char* name) {
-    int ret = PyObject_GetBuffer(
-      pyobj, &buf, PyBUF_FORMAT | PyBUF_STRIDED_RO | PyBUF_ANY_CONTIGUOUS | flags);
-    if (ret < 0) {
-      return ret;
-    }
-    if (dimensions && buf.ndim != dimensions) {
-      PyErr_Format(PyExc_ValueError, "Invalid argument %s: Expected %d dimensions, got %d",
-                   name, dimensions, buf.ndim);
-      PyBuffer_Release(&buf);
-      return -1;
-    }
-    /* We'll get a buffer that's either:
-     * C_CONTIGUOUS (last dimension varies the fastest, i.e., has stride=1) or
-     * F_CONTIGUOUS (first dimension varies the fastest, i.e., has stride=1).
-     * The latter is preferred, since it's already in the format that Halide
-     * needs. It can can be achieved in numpy by passing order='F' during array
-     * creation. However, if we do get a C_CONTIGUOUS buffer, flip the dimensions
-     * (transpose) so we can process it without having to reallocate.
-     */
-    int i, j, j_step;
-    if (PyBuffer_IsContiguous(&buf, 'F')) {
-      j = 0;
-      j_step = 1;
-    } else if (PyBuffer_IsContiguous(&buf, 'C')) {
-      j = buf.ndim - 1;
-      j_step = -1;
-    } else {
-      /* Python checks all dimensions and strides, so this typically indicates
-       * a bug in the array's buffer protocol. */
-      PyErr_Format(PyExc_ValueError, "Invalid buffer: neither C nor Fortran contiguous");
-      PyBuffer_Release(&buf);
-      return -1;
-    }
-    for (i = 0; i < buf.ndim; ++i, j += j_step) {
-        dim[i].min = 0;
-        dim[i].stride = (int)(buf.strides[j] / buf.itemsize); // strides is in bytes
-        dim[i].extent = (int)buf.shape[j];
-        dim[i].flags = 0;
-        if (buf.suboffsets && buf.suboffsets[i] >= 0) {
-            // Halide doesn't support arrays of pointers. But we should never see this
-            // anyway, since we specified PyBUF_STRIDED.
-            PyErr_Format(PyExc_ValueError, "Invalid buffer: suboffsets not supported");
-            PyBuffer_Release(&buf);
-            return -1;
+template<int dimensions>
+struct PyHalideBuffer {
+    static constexpr int dims_to_allocate = (dimensions < 1) ? 1 : dimensions;
+
+    Py_buffer py_buf;
+    halide_dimension_t halide_dim[dims_to_allocate];  // Must allocate at least 1, even if d=0
```

Comment:
Nit: this comment seems like it should be on line 135

Reply:
Done